docs/_docs/development-choices.md

```diff
@@ -0,0 +1,212 @@
+
+While we were furthering our knowledge to better approach the development of the project we stumbled
+upon a very interesting [talk](https://www.youtube.com/watch?v=2JB1_e5wZmU) by Scott Wlaschin and
+later read the book [_"Domain Modelling Made Functional"_](https://pragprog.com/titles/swdddf/domain-modeling-made-functional/).
+This sparked a keen interest in the topic and convinced us to carry out the project by embracing a
+fully functional approach.
+
+All domain concepts are modelled using bare-bone enums and case classes; every action dealing with
+these entities is either defined as simple functions or as extension methods.
+This allowed us to have _self-contained, clutter-free_ and _simple_ files containing all the
+definitions of the main domain concepts; this way, we were able to skim through these files along
+with the domain experts to get precious feedback we could easily use to rework our ubiquitous
+language on the spot.
+
+### Action modelling through monads
+
+All core domain actions take advantage of a _monadic encoding of side effects,_ ranging from failure
+with an exception, to emitting events to reading an immutable global state.
+
+Using monads to model side effects proved useful in three distinct ways:
+
+- The core logic of the resulting code is pretty _easy to read and understand:_ complex actions are
+  modelled using a DSL for effects and can be easily composed together.
+  Moreover, these functions expose the side effects they can perform in their type signature making
+  it _impossible for the programmer to forget to handle them_
+- By reifying side effects as data we were able to _easily test_ the core behaviour of our system
+- This approach is a _good fit to implement the hexagonal architecture:_ all the core logic is pure
+  and does not perform any kind of IO
+  ([_dependency rejection_](https://blog.ploeh.dk/2017/01/27/from-dependency-injection-to-dependency-rejection/))
+  while all necessary dependencies are injected as simple parameters or in the reader monad.
+  Following this discipline makes a hexagonal architecture emerge quite naturally
+
+> A code example from our codebase:
+>
+> ```scala
+> def labelProduct[M[_]: CanRaise[WeightNotInRange]: CanEmit[ProductStocked]: Monad](...): M[LabelledProduct] =
+> for
+>   ...
+>   product <- optionalProduct.ifMissingRaise(WeightNotInRange(...): WeightNotInRange)
+>   labelledProduct = LabelledProduct(product, AvailableQuantity(1), batch.id)
+>   _ <- emit(ProductStocked(labelledProduct): ProductStocked)
+> yield labelledProduct
+> ```
+>
+> Just by reading the type signature one knows that the execution of the function can raise a
+> `WeightNotInRange` error and can emit `ProductStocked` events.
+>
+> `ifMissingRaise` and `emit` are a part of the monadic DSL we devised to write more concise
+> and easy-to-read code: thanks to these functions and Scala's for comprehension we can compose a
+> sequence of small actions to obtain more complex behaviour.
```

Comment:
I'm not sure if I like these big asides

Reply:
How would you change these? I thought they could be a good choice to show some non trivial examples